shared/src/sources/balancer/info_fetching.rs

```diff
@@ -83,10 +95,178 @@ impl PoolInfoFetching for PoolInfoFetcher {
         Ok(WeightedPoolInfo {
-            pool_id,
-            tokens,
+            common: CommonPoolInfo {
+                pool_id,
+                tokens,
+                scaling_exponents,
+            },
             weights,
-            scaling_exponents,
         })
     }
+
+    async fn get_stable_pool_data(&self, pool_address: H160) -> Result<StablePoolInfo> {
+        let mut batch = CallBatch::new(self.web3.transport());
+        let pool_contract = BalancerV2StablePool::at(&self.web3, pool_address);
+        // Need vault and pool_id before we can fetch tokens.
+        let vault = BalancerV2Vault::deployed(&self.web3).await?;
+        let pool_id = H256::from(pool_contract.methods().get_pool_id().call().await?.0);
+
+        // token_data and weight calls can be batched
+        let token_data = vault
+            .methods()
+            .get_pool_tokens(Bytes(pool_id.0))
+            .batch_call(&mut batch);
+        let amplification_parameter = pool_contract
+            .methods()
+            .get_amplification_parameter()
+            .batch_call(&mut batch);
+        batch.execute_all(MAX_BATCH_SIZE).await;
+
+        let tokens = token_data.await?.0;
+
+        let token_decimals = self.token_info_fetcher.get_token_infos(&tokens).await;
+        let ordered_decimals = tokens
+            .iter()
+            .map(|token| token_decimals.get(token).and_then(|t| t.decimals))
+            .collect::<Option<Vec<_>>>()
+            .ok_or_else(|| anyhow!("all token decimals required to build scaling factors"))?;
+        // Note that balancer does not support tokens with more than 18 decimals
+        // https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce70f7663e0ac94b25ed60cb86faaa8199fd9e13/pkg/pool-utils/contracts/BasePool.sol#L497-L508
+        let scaling_exponents = ordered_decimals
+            .iter()
+            .map(|decimals| 18u8.checked_sub(*decimals))
+            .collect::<Option<Vec<_>>>()
+            .ok_or_else(|| anyhow!("token with more than 18 decimals"))?;
+
+        Ok(StablePoolInfo {
+            common: CommonPoolInfo {
+                pool_id,
+                tokens,
+                scaling_exponents,
+            },
+            amplification_parameter: amplification_parameter.await?.0,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::token_info::{MockTokenInfoFetching, TokenInfo};
+    use ethcontract_mock::Mock;
+    use maplit::hashmap;
+
+    #[tokio::test]
+    async fn get_stable_pool_data_err() {
+        let mock = Mock::new(1);
+        let web3 = mock.web3();
+
+        let vault = mock.deploy(BalancerV2Vault::raw_contract().abi.clone());
+        let stable_pool = mock.deploy(BalancerV2StablePool::raw_contract().abi.clone());
+
+        let pool_id = H256::from_low_u64_be(1);
+        let token = H160::from_low_u64_be(1);
+        stable_pool
+            .expect(BalancerV2StablePool::signatures().get_pool_id())
```

Comment:
Nit: prefer `expect_call` or `expect_transaction` to make sure the code uses `call` or `send` as intended.

P.S. maybe we need to rename `expect` to make it clearer? Something like `expect_call_or_transaction`?